packages/hooks/src/index.ts

```diff
@@ -1,32 +1,42 @@
 import { functionHooks } from './function';
-import { HookSettings } from './base';
+import { Middleware } from './compose';
 import { objectHooks, HookMap } from './object';
 import { hookDecorator } from './decorator';
+import { HookManager, HookContextData, HookContext, HookContextConstructor, HookOptions } from './base';
 
 export * from './function';
 export * from './compose';
 export * from './base';
 
-export interface WrapperAddon<F, T = any> {
+export interface WrapperAddon<F> {
   original: F;
-  params: (...params: (string | [string, any])[]) => F&((...rest: any[]) => Promise<T>);
+  Context: HookContextConstructor;
+  createContext: (data?: HookContextData) => HookContext;
 }
 
-// hooks(fn, hookSettings)
+export type WrappedFunction<F, T> = F&((...rest: any[]) => Promise<T>|Promise<HookContext>)&WrapperAddon<F>;
+
+export function middleware (mw: Middleware[]) {
+  const manager = new HookManager();
+
+  return manager.middleware(mw);
+}
+
+// hooks(fn, hookOptions)
 export function hooks<F, T = any> (
-  fn: F, hooks: HookSettings
-): F&((...rest: any[]) => Promise<T>)&WrapperAddon<F, T>;
+  fn: F, manager: HookManager
+): WrappedFunction<F, T>;
 // hooks(object, hookMap)
-export function hooks<O> (obj: O, hookMap: HookMap): O;
-// @hooks(hookSettings)
+export function hooks<O> (obj: O, hookMap: HookMap|Middleware[]): O;
+// @hooks(hookOptions)
 export function hooks<T = any> (
-  hooks?: HookSettings
+  _manager?: HookOptions
 ): any;
 // Fallthrough to actual implementation
 export function hooks (...args: any[]) {
   const [ target, _hooks ] = args;
 
-  if (typeof target === 'function' && Array.isArray(_hooks.middleware || _hooks)) {
+  if (typeof target === 'function' && _hooks instanceof HookManager) {
```

Comment:
```suggestion
  if (typeof target === 'function' && (_hooks instanceof HookManager || Array.isArray(_hooks))) {
```

In tests there is not one that accepts an array of hooks, without being wrapped beforehand:

```js
const sayHi = hooks(hello, [
  async (context, next) => {
    await next();
    context.result += '!';
  }
]);
```